File: src/utils/restoreMode.ts
/**
 * Startup session-restore mode: how the previously-open tabs are handled when
 * the app launches.
 *
 * - `"never"` — never restore; start with a fresh empty session.
 * - `"ask"` — show a dialog offering to restore the previous session.
 * - `"always"` — restore the previous session silently.
 */

import { invoke } from "@tauri-apps/api/core";

import type { AppSettings, SavedConnection } from "@/types/connection";
import type { LastSession } from "@/types/lastSession";

/** The three restore modes. */
export type RestoreLastSessionMode = "never" | "ask" | "always";

/**
 * Reachability of a restorable tab's connection target, resolved by an
 * asynchronous probe after the restore dialog opens.
 *
 * - `"reachable"` — the target answered (host port open / serial device present).
 * - `"unreachable"` — the target is down (host unreachable / device offline);
 *   the dialog flags it with a warning icon.
 * - `"unknown"` — not probed or the probe was inconclusive (the default before
 *   the probe resolves, and for targets with nothing meaningful to probe).
 */
export type RestoreReachability = "reachable" | "unreachable" | "unknown";

/**
 * The connection target derived from a stored tab, used to drive the
 * reachability probe. `kind` selects how the probe checks it; `local` and
 * `agent` targets are not network-probed (they resolve to `"unknown"`).
 */
export interface RestoreTabTarget {
  /** How the tab connects, selecting the reachability check to run. */
  kind: "host" | "serial" | "local" | "agent";
  /** Target host for `host` targets (SSH/telnet). */
  host?: string;
  /** Target TCP port for `host` targets. */
  port?: number;
  /** Serial device path for `serial` targets. */
  device?: string;
  /** Remote agent id for `agent` targets. */
  agentId?: string;
}

/** A single restorable tab, described for the restore dialog. */
export interface RestoreTabInfo {
  /** Human-readable tab title. */
  title: string;
  /** Short connection-type label (e.g. "SSH", "Serial", "Local"). */
  typeLabel: string;
  /**
   * The probe target derived from the stored tab. Optional so display-only
   * callers (and tests) need not construct it; {@link summarizeLastSession}
   * always populates it so the reachability probe can run.
   */
  target?: RestoreTabTarget;
  /**
   * Reachability of {@link target}, set asynchronously once the probe resolves.
   * Absent until then (treated as `"unknown"`).
   */
  reachability?: RestoreReachability;
  /**
   * Short human-readable reason shown beside the warning icon when the target is
   * unreachable (e.g. `"device offline"`, `"host unreachable"`).
   */
  unreachableReason?: string;
}

/** Summary of a stored last session for the restore dialog. */
export interface RestorePrompt {
  /** Total number of restorable tabs across all groups. */
  tabCount: number;
  /** Per-tab descriptors for display. */
  tabs: RestoreTabInfo[];
}

/**
 * Resolve the effective restore mode from settings, migrating the legacy
 * boolean `restoreLastSessionOnStartup` when the explicit mode is unset.
 *
 * - explicit {@link AppSettings.restoreLastSessionMode} wins when valid;
 * - otherwise the legacy boolean `=== false` maps to `"never"`;
 * - otherwise the default is `"ask"` (the concept default).
 *
 * The decision logic lives in `core::restore_mode` (Rust); this delegates to the
 * `restore_resolve_mode` command so there is a single source of truth (#2200).
 */
export async function resolveRestoreMode(settings: AppSettings): Promise<RestoreLastSessionMode> {
  return await invoke<RestoreLastSessionMode>("restore_resolve_mode", { settings });
}

/**
 * Flatten a stored {@link LastSession} into a per-tab summary for the restore
 * dialog. Iterates groups → leaves → tabs in a stable order; that same order is
 * the tab index space used by {@link filterSessionBySelection}.
 *
 * `connections` (optional) lets `connectionRef` tabs resolve their host/serial
 * target for the reachability probe; pass the loaded connections when available.
 *
 * The summarisation is pure and runs in `core::restore_mode` (#2200): each tab
 * carries the {@link RestoreTabTarget} the probe needs, but the **asynchronous
 * reachability probe itself stays client-side** ({@link RestoreTabInfo.reachability}
 * is populated afterwards by `probeRestoreTargets`). That is the async-probe
 * seam — pure decision server-side, network I/O on the client.
 */
export async function summarizeLastSession(
  session: LastSession,
  connections: SavedConnection[] = []
): Promise<RestorePrompt> {
  return await invoke<RestorePrompt>("restore_summarize_last_session", { session, connections });
}

/**
 * Prune a stored {@link LastSession} down to the tabs the user chose to restore.
 *
 * `selected` holds the flat tab indices to keep, in the same order
 * {@link summarizeLastSession} produced. Leaves left with no tabs are dropped,
 * splits collapse when only one child survives (redistributing sizes), and tab
 * groups whose layout empties out are removed. `activeGroupIndex` is remapped to
 * the surviving group nearest the original active one; `windows` are preserved
 * untouched (an emptied window still round-trips, per #1902).
 *
 * Delegates to the `restore_filter_session_by_selection` core command (#2200).
 */
export async function filterSessionBySelection(
  session: LastSession,
  selected: ReadonlySet<number>
): Promise<LastSession> {
  return await invoke<LastSession>("restore_filter_session_by_selection", {
    session,
    selected: [...selected],
  });
}

File: src-tauri/src/commands/restore_mode.rs
//! Tauri command surface for the startup session-restore decision logic (#2200).
//!
//! Wires the previously-dormant `termihub_core::restore_mode` port (#2145) into
//! the frontend restore path. Each command is a thin, **stateless** wrapper over
//! one of the core's pure decision functions; the frontend's
//! `src/utils/restoreMode.ts` now delegates its logic here so there is a single
//! source of truth (the Rust port), proven equivalent to the retired TypeScript
//! logic by the #2145 golden vectors (`core/tests/restore_mode_golden.rs`).
//!
//! ## The async-probe seam
//!
//! [`restore_summarize_last_session`] does the **pure** work: it flattens a
//! stored session into per-tab descriptors, each carrying the connection
//! `target` its reachability check would probe. It never performs the probe.
//! The asynchronous reachability probe stays entirely client-side
//! (`src/utils/restoreReachability.ts` + `probeRestorePromptReachability`),
//! consuming the `target`s this command produced and patching `reachability`
//! onto the prompt afterwards. That decoupling — pure decision here, async I/O
//! probe on the client — is why the lift is partial rather than whole.

use std::collections::HashSet;

use termihub_core::restore_mode::{
    filter_session_by_selection, resolve_restore_mode, summarize_last_session, AppSettings,
    LastSession, RestoreLastSessionMode, RestorePrompt, SavedConnection,
};

/// Resolve the effective restore mode from settings, migrating the legacy
/// `restoreLastSessionOnStartup` boolean when the explicit mode is unset.
#[tauri::command]
pub fn restore_resolve_mode(settings: AppSettings) -> RestoreLastSessionMode {
    resolve_restore_mode(&settings)
}

/// Flatten a stored last session into the per-tab restore-dialog summary. Each
/// tab carries the `target` the client-side reachability probe then consumes.
#[tauri::command]
pub fn restore_summarize_last_session(
    session: LastSession,
    connections: Vec<SavedConnection>,
) -> RestorePrompt {
    summarize_last_session(&session, &connections)
}

/// Prune a stored last session down to the flat tab indices the user selected
/// (same order [`restore_summarize_last_session`] produced).
#[tauri::command]
pub fn restore_filter_session_by_selection(
    session: LastSession,
    selected: Vec<i64>,
) -> LastSession {
    let selected: HashSet<i64> = selected.into_iter().collect();
    filter_session_by_selection(&session, &selected)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    // These assert the command wrappers pass their arguments straight through to
    // the core functions unchanged; the behaviour of the functions themselves is
    // proven equivalent to the retired TS logic by `restore_mode_golden.rs`.

    #[test]
    fn resolve_mode_delegates_to_core() {
        let settings: AppSettings =
            serde_json::from_value(json!({ "restoreLastSessionMode": "always" })).unwrap();
        assert_eq!(
            restore_resolve_mode(settings.clone()),
            resolve_restore_mode(&settings)
        );
    }

    #[test]
    fn summarize_delegates_to_core() {
        let session: LastSession = serde_json::from_value(json!({
            "version": "1",
            "activeGroupIndex": 0,
            "tabGroups": [{
                "name": "g",
                "layout": { "type": "leaf", "tabs": [
                    { "title": "shell", "inlineConfig": { "type": "local", "config": {} } }
                ] }
            }]
        }))
        .unwrap();
        assert_eq!(
            restore_summarize_last_session(session.clone(), Vec::new()),
            summarize_last_session(&session, &[])
        );
    }

    #[test]
    fn filter_collects_selection_into_set() {
        let session: LastSession = serde_json::from_value(json!({
            "version": "1",
            "activeGroupIndex": 0,
            "tabGroups": [{
                "name": "g",
                "layout": { "type": "leaf", "tabs": [
                    { "title": "a", "inlineConfig": { "type": "local", "config": {} } },
                    { "title": "b", "inlineConfig": { "type": "local", "config": {} } }
                ] }
            }]
        }))
        .unwrap();
        let expected = filter_session_by_selection(&session, &HashSet::from([0]));
        assert_eq!(
            restore_filter_session_by_selection(session, vec![0]),
            expected
        );
    }
}
